cmake/modules/SetUpLinux.cmake

```diff
@@ -106,18 +106,18 @@ if(CMAKE_COMPILER_IS_GNUCXX)
   set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--no-undefined -Wl,--hash-style=\"both\"")
 
   # Select flags.
-  set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "-O3 -g -DNDEBUG")
-  set(CMAKE_CXX_FLAGS_RELEASE        "-O3 -DNDEBUG")
-  set(CMAKE_CXX_FLAGS_OPTIMIZED      "-Ofast -DNDEBUG")
-  set(CMAKE_CXX_FLAGS_DEBUG          "-g")
-  set(CMAKE_CXX_FLAGS_DEBUGFULL      "-g3")
-  set(CMAKE_CXX_FLAGS_PROFILE        "-g3 -ftest-coverage -fprofile-arcs")
-  set(CMAKE_C_FLAGS_RELWITHDEBINFO   "-O3 -g -DNDEBUG")
-  set(CMAKE_C_FLAGS_RELEASE          "-O3 -DNDEBUG")
-  set(CMAKE_C_FLAGS_OPTIMIZED        "-Ofast -DNDEBUG")
-  set(CMAKE_C_FLAGS_DEBUG            "-g")
-  set(CMAKE_C_FLAGS_DEBUGFULL        "-g3 -fno-inline")
-  set(CMAKE_C_FLAGS_PROFILE          "-g3 -fno-inline -ftest-coverage -fprofile-arcs")
+  set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "-O3 -g -DNDEBUG"  CACHE STRING "Flags for release build with debug info")
+  set(CMAKE_CXX_FLAGS_RELEASE        "-O3 -DNDEBUG"     CACHE STRING "Flags for release build")
+  set(CMAKE_CXX_FLAGS_OPTIMIZED      "-Ofast -DNDEBUG"  CACHE STRING "Flags for a highly optimised build")
+  set(CMAKE_CXX_FLAGS_DEBUG          "-g"               CACHE STRING "Flags for a debug build")
+  set(CMAKE_CXX_FLAGS_DEBUGFULL      "-g3"              CACHE STRING "Flags for a debug build with more debug info")
+  set(CMAKE_CXX_FLAGS_PROFILE        "-g3 -ftest-coverage -fprofile-arcs" CACHE STRING "")
+  set(CMAKE_C_FLAGS_RELWITHDEBINFO   "-O3 -g -DNDEBUG"  CACHE STRING "Flags for release build with debug info")
+  set(CMAKE_C_FLAGS_RELEASE          "-O3 -DNDEBUG"     CACHE STRING "Flags for release build")
+  set(CMAKE_C_FLAGS_OPTIMIZED        "-Ofast -DNDEBUG"  CACHE STRING "Flags for a highly optimised build")
+  set(CMAKE_C_FLAGS_DEBUG            "-g"               CACHE STRING "Flags for a debug build")
+  set(CMAKE_C_FLAGS_DEBUGFULL        "-g3 -fno-inline"  CACHE STRING "Flags for a debug build with more debug info")
+  set(CMAKE_C_FLAGS_PROFILE          "-g3 -fno-inline -ftest-coverage -fprofile-arcs" CACHE STRING "")
```

Comment:
@Axel-Naumann I wonder if we really need all those build types. We could move to use only the usual `Release`, `Debug`, `RelWithDebInfo`, and `MinSizeRel`, which are the defaults provided by CMake. Maybe we can leave these settings as CMake's defaults, which should be reasonable enough, and set [`CMAKE_<LANG>_FLAGS_INIT`](https://cmake.org/cmake/help/v3.9/variable/CMAKE_LANG_FLAGS_INIT.html) and friends if needed.

Reply:
That's not my business. I don't know who added them in the first place ...

Reply:
I don't have a strong opinion, we should do what you recommend. But that can probably be done in a subsequent PR?

Reply:
Yeah, go ahead, we can continue in subsequent PRs.